### PR TITLE
Add a command-line tool for asset compression

### PR DESCRIPTION
**Pull request by jcelerier**
The command-line API is fairly simple for now (it only allows to set the overall quantization level) - I think that as @seanharmer mentioned the correct thing to do would be to have a project file to indicate the whole set of compression parameters.

### CODE REVIEW

File: tools/assetcompressor/main.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    QCoreApplication app(argc, argv);
+    app.setApplicationName(QStringLiteral("Kuesa Asset Compressor"));
+    app.setOrganizationDomain(QStringLiteral("kdab.com"));
+    app.setOrganizationName(QStringLiteral("KDAB"));
+    app.setApplicationVersion(QStringLiteral("1.0"));
+
+    QCommandLineParser cmdline;
+    cmdline.addHelpOption();
+    cmdline.addVersionOption();
+    cmdline.addPositionalArgument("input", QObject::tr("glTF2 file to compress"));
```
Comment (jjcasmar) [inline tools/assetcompressor/main.cpp]: should we add -i and -o options?

File: tools/assetcompressor/main.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        return 1;
+    }
+
+    int ratio = cmdline.value(compRatio).toInt();
+    if (ratio < 1 || ratio > 16) {
+        qCritical("Invalid compression ratio.");
+        return 1;
+    }
+
+    QString asset = args[0];
```
Comment (lemirep) [inline tools/assetcompressor/main.cpp]: const QString

File: tools/assetcompressor/main.cpp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    cmdline.process(app);
+
+    const auto &args = cmdline.positionalArguments();
+    if (args.size() != 2) {
+        qCritical("Invalid arguments specified.\n%s", cmdline.helpText().toLatin1().constData());
+        return 1;
+    }
+
+    int ratio = cmdline.value(compRatio).toInt();
```
Comment (lemirep) [inline tools/assetcompressor/main.cpp]: const int

File: tools/assetcompressor/main.cpp [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        return 1;
+    }
+
+    QString asset = args[0];
+    if (!QFileInfo::exists(asset)) {
+        qCritical("Could not find input file.");
+        return 1;
+    }
+
+    QDir sourceDir = QFileInfo(asset).dir();
```
Comment (lemirep) [inline tools/assetcompressor/main.cpp]: const

File: tools/assetcompressor/main.cpp [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    // Parsing
+    SceneEntity scene;
+    GLTF2Context ctx;
+
+    GLTF2Parser parser(&scene);
+    parser.setContext(&ctx);
+
+    auto res = parser.parse(asset);
```
Comment (lemirep) [inline tools/assetcompressor/main.cpp]: const bool I suppose

File: tools/assetcompressor/main.cpp [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    GLTF2Exporter exporter;
+    exporter.setContext(&ctx);
+    exporter.setScene(&scene);
+    exporter.setConfiguration(configuration);
+
+    QElapsedTimer timer;
+    timer.start();
+
+    auto new_asset = exporter.saveInFolder(sourceDir, targetDir);
```
Comment (lemirep) [inline tools/assetcompressor/main.cpp]: Do we even care about new_asset here? I don't see it being used below

File: tools/assetcompressor/main.cpp (hunk 5 above)
Comment (er-vin) [reply tools/assetcompressor/main.cpp]: Or const auto.

You live so much in 1998. ;-)

File: tools/assetcompressor/main.cpp (hunk 1 above)
Comment (er-vin) [reply tools/assetcompressor/main.cpp]: Agreed

File: tools/assetcompressor/main.cpp (hunk 3 above)
Comment (er-vin) [reply tools/assetcompressor/main.cpp]: or const auto

File: tools/assetcompressor/main.cpp (hunk 2 above)
Comment (er-vin) [reply tools/assetcompressor/main.cpp]: or const auto

File: tools/assetcompressor/main.cpp (hunk 6 above)
Comment (er-vin) [reply tools/assetcompressor/main.cpp]: Indeed.

File: tools/assetcompressor/main.cpp (hunk 1 above)
Comment (jcelerier) [reply tools/assetcompressor/main.cpp]: done

File: tools/assetcompressor/main.cpp (hunk 2 above)
Comment (jcelerier) [reply tools/assetcompressor/main.cpp]: done

File: tools/assetcompressor/main.cpp (hunk 3 above)
Comment (jcelerier) [reply tools/assetcompressor/main.cpp]: done

File: tools/assetcompressor/main.cpp (hunk 4 above)
Comment (jcelerier) [reply tools/assetcompressor/main.cpp]: done

File: tools/assetcompressor/main.cpp (hunk 5 above)
Comment (jcelerier) [reply tools/assetcompressor/main.cpp]: done

File: tools/assetcompressor/main.cpp (hunk 6 above)
Comment (jcelerier) [reply tools/assetcompressor/main.cpp]: that was a bug :)

**Comment by jcelerier** [comment]
I have a remaining problem that will need another PR before merging : to compute the statistics it looks for "compressedBuffer.bin" but the actual name may be "compressedBuffer-1.bin", "compressedBuffer-2.bin", ...

So the exporter class should have a way to retrieve the name of the file it generated for this.

File: tools/assetcompressor/assetcompressor.pro [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+TEMPLATE = app
+
+QT += kuesa kuesa-private
```
Comment (mkrus) [inline tools/assetcompressor/assetcompressor.pro]: remove gui

File: tools/assetcompressor/main.cpp [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    exporter.setScene(&scene);
+    exporter.setConfiguration(configuration);
+
+    const auto sourceDir = assetFileInfo.dir();
+
+    // Meaasure the actual export step duration
+    QElapsedTimer timer;
+    timer.start();
+
+    const auto new_asset = exporter.saveInFolder(sourceDir, targetDir);
```
Comment (mkrus) [inline tools/assetcompressor/main.cpp]: should this fail is any of the output files exists unless a `-f|--force` option is added?

File: tools/assetcompressor/assetcompressor.pro (hunk 7 above)
Comment (jcelerier) [reply tools/assetcompressor/assetcompressor.pro]: done

File: tools/assetcompressor/main.cpp (hunk 8 above)
Comment (jcelerier) [reply tools/assetcompressor/main.cpp]: that would be safer indeed, I will have to finish https://github.com/KDAB/kuesa/pull/49 first

**Comment by jcelerier** [comment]
> should it be called `assetprocessor`?

makes sense to me

File: tools/assetcompressor/main.cpp (hunk 8 above)
Comment (jcelerier) [reply tools/assetcompressor/main.cpp]: done

**Pull request merged**